src/controller/controller.lua

```diff
@@ -0,0 +1,58 @@
+local Timer = require("timer/timer")
+local utils = require("utils/utils")
+local settings = require("settings/settings")
+
+local Controller = {}
+
+function Controller:new(board)
+    local controller = {
+        board = board,
+        timers = {
+            ['moveLeft'] = Timer:new(),
+            ['moveRight'] = Timer:new(),
+            ['moveDown'] = Timer:new(),
+        },
+    }
+    self.__index = self
+    setmetatable(controller, self)
+    return controller
+end
+
+function Controller:run(action)
```

Comment:
if action == 'left' then self.board.movement:moveLeft() end
    if action == 'right' then self.board.movement:moveRight() end
    if action == 'down' then self.board.movement:moveDown() end
    if action == 'up' then self.board.movement:rotateCw() end
    if action == 'z' then self.board.movement:rotateCcw() end
    if action == 'space' then self.board.movement:hardDrop() end
    if action == 'lshift' then self.board:holdCurrentTetromino() end
    if action == 'rshift' then self.board:holdCurrentTetromino() end
    if action == 'c' then self.board:holdCurrentTetromino() end

Reply:
So an even better way is somehow making a table of functions but I found it hard to do with the OOP and passing in "self". If you are able to then you can save some performance.

handleNonRepeatKeys would then just be `self.table[key](self)` and handleRepeatKeys would use one less for loop.

If you can figure it out then awesome, if not that is okay.
http://lua-users.org/wiki/SwitchStatement

Reply:
Also replace the ' with "

Reply:
Gonna try something like this: https://stackoverflow.com/questions/45016391/lua-strange-behavior-with-loadstring

Reply:
Also the problem with setting the actions == key is that it hard codes the key bindings, for example the Z key has to be rotate CCW.

Reply:
Okay keep the action as before but implement changing the order or key and value in settings.